src/backend/distributed/master/master_modify_multiple_shards.c

```diff
@@ -0,0 +1,403 @@
+			Assert(shardConnectionsFound);
+
+			deparse_shard_query(query, relationId, shardId, shardQueryString);
+			shardQueryStringData = shardQueryString->data;
+			shardAffectedTupleCount = SendQueryToConnections(shardQueryStringData,
+															 shardConnections);
+			affectedTupleCount += shardAffectedTupleCount;
+		}
+
+		if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
+		{
+			PrepareRemoteTransactions(allShardsConnectionList);
+		}
+	}
+	PG_CATCH();
+	{
+		/* roll back all transactions */
+		AbortRemoteTransactions(allShardsConnectionList);
+		CloseConnections(allShardsConnectionList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	CommitRemoteTransactions(allShardsConnectionList, false);
+	CloseConnections(allShardsConnectionList);
+
+	return affectedTupleCount;
+}
+
+
+/**
+ * OpenConnectionsToAllShardPlacement opens connections to all placements of
+ * the given shard list and returns the hash table containing the connections.
+ * The resulting hash table maps shardId to ShardConnection struct.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIntervalList)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+
+	ListCell *shardIntervalCell = NULL;
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash);
+	}
+
+	return shardConnectionHash;
+}
+
+
+/**
+ * OpenConnectionsToShardPlacements opens connections to all placements of the
+ * shard with the given shardId and populates the shardConnectionHash table
+ * accordingly.
+ */
+static void
+OpenConnectionsToShardPlacements(uint64 shardId, HTAB *shardConnectionHash)
+{
+	bool shardConnectionsFound = false;
+
+	/* get existing connections to the shard placements, if any */
+	ShardConnections *shardConnections = GetShardConnections(shardConnectionHash,
+															 shardId,
+															 &shardConnectionsFound);
+
+	List *shardPlacementList = FinalizedShardPlacementList(shardId);
+	ListCell *shardPlacementCell = NULL;
+	List *connectionList = NIL;
+
+	Assert(!shardConnectionsFound);
+
+	if (shardPlacementList == NIL)
+	{
+		ereport(ERROR, (errmsg("Could not find any shard placements for the shard "
+							   UINT64_FORMAT, shardId)));
+	}
+
+	foreach(shardPlacementCell, shardPlacementList)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+			shardPlacementCell);
+		char *workerName = shardPlacement->nodeName;
+		uint32 workerPort = shardPlacement->nodePort;
+		char *nodeUser = CurrentUserName();
+		PGconn *connection = ConnectToNode(workerName, workerPort, nodeUser);
+		TransactionConnection *transactionConnection = NULL;
+
+		if (connection == NULL)
+		{
+			List *abortConnectionList = ConnectionList(shardConnectionHash);
+			CloseConnections(abortConnectionList);
+
+			ereport(ERROR, (errmsg("Could not establish a connection to all "
```

Comment:
as a PostgreSQL convention, errmsg messsages start with a small letter and have no . at the end.